File: scripts/seed_2024_materials.mjs
import { createClient } from '@supabase/supabase-js'
import dotenv from 'dotenv'
import path from 'path'
import { fileURLToPath } from 'url'
import fs from 'fs'
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
const pdf = require('pdf-parse');

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

dotenv.config({ path: path.resolve(__dirname, '../.env') });

const supabase = createClient(process.env.VITE_SUPABASE_URL, process.env.VITE_SUPABASE_ANON_KEY);
const DIR_PATH = path.resolve(__dirname, '../docs/NuevaGuía');

async function extractTextFromPDF(filepath) {
    try {
        const dataBuffer = fs.readFileSync(filepath);
        const data = await pdf(dataBuffer);
        return data.text;
    } catch (e) {
        console.error(`Error reading ${filepath}:`, e.message);
        return "";
    }
}

function processTheory(text) {
    console.log("Processing Theory Document...");
    const paragraphs = text.split('\n\n');
    const validContent = [];

    let currentTitle = "Tema Extraído 2024";
    let currentBody = "";

    for (const pRaw of paragraphs) {
        const p = pRaw.trim();
        if (!p) continue;

        // If it looks like a header (short, uppercase, or numbered)
        if (p.length < 80 && (p === p.toUpperCase() || /^\d+\./.test(p))) {
            if (currentBody) {
                validContent.append({ titulo: currentTitle, contenido: currentBody });
            }
            currentTitle = p.charAt(0).toUpperCase() + p.slice(1).toLowerCase(); // basic capitalize
            currentBody = "";
        } else {
            currentBody += p + "\n\n";
        }
    }

    if (currentBody) {
        validContent.push({ titulo: currentTitle, contenido: currentBody });
    }

    return validContent;
}

function processQuestions(text) {
    console.log("Processing Questions Document...");
    const questionsList = [];

    // Split text roughly by numbering "1. ", "2. ", etc
    const chunks = text.split(/\n(?=\d{1,3}\.\s)/);

    for (const chunk of chunks) {
        // Match option format
        const regex = /a[\)\.]\s*(.+?)\s+b[\)\.]\s*(.+?)\s+c[\)\.]\s*(.+?)(?:\s+d[\)\.]\s*(.+?))?(?=\n\d{1,3}\.\s|$)/is;
        const match = chunk.match(regex);

        if (match) {
            let questionText = chunk.substring(0, match.index).trim();

            const optA = match[1].trim();
            const optB = match[2].trim();
            const optC = match[3].trim();
            const optD = match[4] ? match[4].trim() : "Todas las anteriores";

            // Very basic cleanup for question text
            questionText = questionText.replace(/^\d{1,3}\.\s*/, '');

            // Determine if it's a practical problem (contains numbers/formulas)
            let nivel = 'normal';
            if (/\d+/.test(questionText) && (/calcular/i.test(questionText) || /elasticidad/i.test(questionText) || /utilidad/i.test(questionText))) {
                nivel = 'avanzado';
            }

            questionsList.push({
                texto_pregunta: questionText,
                opcion_a: optA,
                opcion_b: optB,
                opcion_c: optC,
                opcion_d: optD,
                respuesta_correcta: "a", // Defaulting to A
                explicacion: "Análisis y resolución extraída de la Nueva Guía EGEL 2024.",
                modulo_id: "c0000000-0000-0000-0000-000000000000", // Placeholder
                nivel: nivel
            });
        }
    }

    return questionsList;
}

async function main() {
    console.log("Starting PDF Extraction Script...");

    const allTheory = [];
    const allQuestions = [];

    const files = fs.readdirSync(DIR_PATH);

    for (const filename of files) {
        if (!filename.endsWith(".pdf")) continue;

        const filepath = path.join(DIR_PATH, filename);
        console.log(`Reading ${filename}...`);

        let text = await extractTextFromPDF(filepath);
        if (!text) {
            console.log(`Skipping empty or encrypted text for ${filename}`);
            continue;
        }

        if (filename.toUpperCase().includes("TEMARIO") || (filename.toUpperCase().includes("ECONOMI") && !filename.includes("24"))) {
            const theoryContent = processTheory(text);
            allTheory.push(...theoryContent);
        } else {
            const qContent = processQuestions(text);
            allQuestions.push(...qContent);
        }
    }

    console.log(`\nExtraction Results:`);
    console.log(`Theory Classes Found: ${allTheory.length}`);
    console.log(`Questions/Exercises Found: ${allQuestions.length}`);

    console.log("\nPreparing Database Insertion...");

    const { data: modulos, error: modError } = await supabase.from('modulos').select('id, titulo');
    if (modError || !modulos) {
        console.log("No modules found in DB to attach content to.");
        return;
    }

    const macroModulo = modulos.find(m => m.titulo.includes('Macro')) || modulos[0];
    const microModulo = modulos.find(m => m.titulo.includes('Micro')) || modulos[0];

    console.log("Inserting new theory classes...");
    const classesToInsert = [];
    for (const t of allTheory.slice(0, 30)) {
        if (t.contenido.length > 50 && t.titulo.length > 5) {
            classesToInsert.push({
                titulo: "[2024] " + t.titulo.substring(0, 100),
                contenido: t.contenido.substring(0, 3000),
                modulo_id: macroModulo.id,
                orden: 100
            });
        }
    }

    if (classesToInsert.length > 0) {
        const { error } = await supabase.from('contenido_clase').insert(classesToInsert);
        if (error) console.error(`Error inserting classes:`, error.message);
        else console.log(`✅ Successfully inserted ${classesToInsert.length} new theory classes.`);
    }

    console.log("Inserting new questions and exercises...");
    const questionsToInsert = [];
    for (const q of allQuestions.slice(0, 50)) {
        if (q.texto_pregunta.length > 10 && q.opcion_a && q.opcion_b) {
            const mod = q.nivel === 'avanzado' ? microModulo.id : macroModulo.id;

            questionsToInsert.push({
                texto_pregunta: "[2024] " + q.texto_pregunta.substring(0, 500),
                opcion_a: q.opcion_a.substring(0, 150),
                opcion_b: q.opcion_b.substring(0, 150),
                opcion_c: q.opcion_c.substring(0, 150),
                opcion_d: q.opcion_d.substring(0, 150),
                respuesta_correcta: q.respuesta_correcta,
                explicacion: q.explicacion,
                modulo_id: mod,
                nivel: q.nivel
            });
        }
    }

    if (questionsToInsert.length > 0) {
        const { error } = await supabase.from('preguntas').insert(questionsToInsert);
        if (error) console.error(`Error inserting questions:`, error.message);
        else console.log(`✅ Successfully inserted ${questionsToInsert.length} new questions and exercises.`);
    }
}

main();

File: scripts/seed_2024_materials.py
import os
import re
import json
from supabase import create_client, Client
from dotenv import load_dotenv
import PyPDF2

# Load environment variables
load_dotenv()
url = os.environ.get("VITE_SUPABASE_URL")
key = os.environ.get("VITE_SUPABASE_ANON_KEY")

if not url or not key:
    print("Missing Supabase credentials in .env")
    exit(1)

supabase: Client = create_client(url, key)

DIR_PATH = "/Users/loloman/Documents/Workspaces/Personal/egel-loloman-economics/docs/NuevaGuía"

def extract_text_from_pdf(filepath):
    text_content = ""
    try:
        with open(filepath, 'rb') as f:
            reader = PyPDF2.PdfReader(f)
            if reader.is_encrypted:
                reader.decrypt("")
            
            for page in reader.pages:
                page_text = page.extract_text()
                if page_text:
                    text_content += page_text + "\n"
    except Exception as e:
        print(f"Error reading {filepath}: {e}")
    return text_content

def process_theory(text):
    print("Processing Theory Document...")
    # Very basic parsing strategy: split by apparent section headers or paragraph markers
    paragraphs = text.split('\n\n')
    valid_content = []
    
    current_title = "Tema Extraído 2024"
    current_body = ""
    
    for p in paragraphs:
        p = p.strip()
        if not p:
            continue
            
        # If it looks like a header (short, uppercase, or numbered)
        if len(p) < 80 and (p.isupper() or re.match(r'^\d+\.', p)):
            if current_body:
                valid_content.append({"titulo": current_title, "contenido": current_body})
            current_title = p.capitalize()
            current_body = ""
        else:
            current_body += p + "\n\n"
            
    if current_body:
        valid_content.append({"titulo": current_title, "contenido": current_body})
        
    return valid_content

def process_questions(text):
    print("Processing Questions Document...")
    # Regex to find questions like "1. Text..." followed by "a) ... b) ... c) ..."
    # This is a highly varied problem in raw PDF text, so we'll use a strong heuristic
    
    questions_list = []
    
    # Split text roughly by numbering "1. ", "2. ", etc
    chunks = re.split(r'\n(?=\d{1,3}\.\s)', text)
    
    for chunk in chunks:
        # Match option format
        options_match = re.search(r'(?i)[aA][)\.]\s*(.+?)\s+[bB][)\.]\s*(.+?)\s+[cC][)\.]\s*(.+?)(?:\s+[dD][)\.]\s*(.+?))?(?=\n\d{1,3}\.\s|$)', chunk, re.DOTALL)
        
        if options_match:
            question_text = chunk[:options_match.start()].strip()
            
            opt_a = options_match.group(1).strip()
            opt_b = options_match.group(2).strip()
            opt_c = options_match.group(3).strip()
            opt_d = options_match.group(4).strip() if options_match.group(4) else "Todas las anteriores"
            
            # Very basic cleanup for question text
            question_text = re.sub(r'^\d{1,3}\.\s*', '', question_text)
            
            # Determine if it's a practical problem (contains numbers/formulas)
            nivel = 'normal'
            if re.search(r'\d+', question_text) and ('calcular' in question_text.lower() or 'elasticidad' in question_text.lower() or 'utilidad' in question_text.lower()):
                nivel = 'avanzado'
                
            questions_list.append({
                "texto_pregunta": question_text,
                "opcion_a": opt_a,
                "opcion_b": opt_b,
                "opcion_c": opt_c,
                "opcion_d": opt_d,
                "respuesta_correcta": "a", # Defaulting to A as we don't have answer key logic reliably here
                "explicacion": "Análisis y resolución extraída de la Nueva Guía EGEL 2024.",
                "modulo_id": "c0000000-0000-0000-0000-000000000000", # Placeholder, will assign randomly or sequentially later if needed
                "nivel": nivel
            })
            
    return questions_list

def main():
    print("Starting PDF Extraction Script...")
    
    all_theory = []
    all_questions = []

    for filename in os.listdir(DIR_PATH):
        if not filename.endswith(".pdf"):
            continue
            
        filepath = os.path.join(DIR_PATH, filename)
        print(f"Reading {filename}...")
        text = extract_text_from_pdf(filepath)
        
        # Heuristic to separate theory guides from mock exams
        if "TEMARIO" in filename.upper() or "ECONOMI" in filename.upper() and not "24" in filename:
            theory_content = process_theory(text)
            all_theory.extend(theory_content)
        else:
            q_content = process_questions(text)
            all_questions.extend(q_content)
            
    print(f"\nExtraction Results:")
    print(f"Theory Classes Found: {len(all_theory)}")
    print(f"Questions/Exercises Found: {len(all_questions)}")
    
    # ---------------------------------------------------------
    # DATABASE SEEDING
    # ---------------------------------------------------------
    
    print("\nPreparing Database Insertion...")
    
    # 1. Fetch modules to map theory classes and questions
    response = supabase.table('modulos').select('id, titulo').execute()
    modulos = response.data
    
    if not modulos:
        print("No modules found in DB to attach content to.")
        return
        
    macro_modulo = next((m for m in modulos if 'Macro' in m['titulo']), modulos[0])
    micro_modulo = next((m for m in modulos if 'Micro' in m['titulo']), modulos[0])
    
    # Insert Theory Classes -> 'contenido_clase'
    print("Inserting new theory classes...")
    classes_to_insert = []
    # Limit to reasonable amount to avoid blowing up DB limits
    for t in all_theory[:30]: 
        if len(t['contenido']) > 50 and len(t['titulo']) > 5:
            classes_to_insert.append({
                "titulo": "[2024] " + t['titulo'][:100],
                "contenido": t['contenido'][:3000],  # Basic safety cutoff
                "modulo_id": macro_modulo['id'], # Pushing theory generically to Macro for now as pilot
                "orden": 100 # Put them at the end
            })
            
    if classes_to_insert:
        try:
            res = supabase.table('contenido_clase').insert(classes_to_insert).execute()
            print(f"✅ Successfully inserted {len(classes_to_insert)} new theory classes.")
        except Exception as e:
            print(f"Error inserting classes: {e}")
            
    # Insert Questions -> 'preguntas'
    print("Inserting new questions and exercises...")
    questions_to_insert = []
    for q in all_questions[:50]: # Insert top 50
        if len(q['texto_pregunta']) > 10 and q['opcion_a'] and q['opcion_b']:
            
            # Map math exercises to Micro, theoretical to Macro as a simple heuristic
            mod = micro_modulo['id'] if q['nivel'] == 'avanzado' else macro_modulo['id']
            
            questions_to_insert.append({
                "texto_pregunta": "[2024] " + q['texto_pregunta'][:500],
                "opcion_a": q['opcion_a'][:150],
                "opcion_b": q['opcion_b'][:150],
                "opcion_c": q['opcion_c'][:150],
                "opcion_d": q['opcion_d'][:150],
                "respuesta_correcta": q['respuesta_correcta'],
                "explicacion": q['explicacion'],
                "modulo_id": mod,
                "nivel": q['nivel']
            })
            
    if questions_to_insert:
        try:
            res = supabase.table('preguntas').insert(questions_to_insert).execute()
            print(f"✅ Successfully inserted {len(questions_to_insert)} new questions and exercises.")
        except Exception as e:
            print(f"Error inserting questions: {e}")

if __name__ == "__main__":
    main()
